external/esp_idf_port/esp32/spi_ram/spiram_psram.c

```diff
@@ -0,0 +1,723 @@
+#include "spi_common.h"
+#include "periph_ctrl.h"
+
+#if CONFIG_SPIRAM_SUPPORT
+#include "soc/rtc.h"
+
+//Commands for PSRAM chip
+#define PSRAM_READ              0x03
+#define PSRAM_FAST_READ         0x0B
+#define PSRAM_FAST_READ_DUMMY   0x3
+#define PSRAM_FAST_READ_QUAD    0xEB
+#define PSRAM_WRITE             0x02
+#define PSRAM_QUAD_WRITE        0x38
+#define PSRAM_ENTER_QMODE       0x35
+#define PSRAM_EXIT_QMODE        0xF5
+#define PSRAM_RESET_EN          0x66
+#define PSRAM_RESET             0x99
+#define PSRAM_SET_BURST_LEN     0xC0
+#define PSRAM_DEVICE_ID         0x9F
+
+#if CONFIG_SPIRAM_TYPE_ESPPSRAM32
+
+#define PSRAM_MFG_ID_M          0xff
+#define PSRAM_MFG_ID_S             8
+#define PSRAM_MFG_ID_V          0x5d
+
+#endif
+
+// IO-pins for PSRAM. These need to be in the VDD_SIO power domain because all chips we
+// currently support are 1.8V parts.
+// WARNING: PSRAM shares all but the CS and CLK pins with the flash, so these defines
+// hardcode the flash pins as well, making this code incompatible with either a setup
+// that has the flash on non-standard pins or ESP32s with built-in flash.
+#define FLASH_CLK_IO      6  //Psram clock is a delayed version of this in 40MHz mode
+#define FLASH_CS_IO       11
+#define PSRAM_CLK_IO      17
+#define PSRAM_CS_IO       16
+#define PSRAM_SPIQ_IO     7
+#define PSRAM_SPID_IO     8
+#define PSRAM_SPIWP_IO    10
+#define PSRAM_SPIHD_IO    9
+
+#define PSRAM_INTERNAL_IO_28       28
+#define PSRAM_INTERNAL_IO_29       29
+#define PSRAM_IO_MATRIX_DUMMY_40M   1
+#define PSRAM_IO_MATRIX_DUMMY_80M   2
+
+#define _SPI_CACHE_PORT   0
+#define _SPI_FLASH_PORT   1
+#define _SPI_80M_CLK_DIV  1
+#define _SPI_40M_CLK_DIV  2
+
+static const char* TAG = "psram";
+typedef enum {
+    PSRAM_SPI_1  = 0x1,
+    PSRAM_SPI_2,
+    PSRAM_SPI_3,
+    PSRAM_SPI_MAX ,
+} psram_spi_num_t;
+
+static psram_cache_mode_t s_psram_mode = PSRAM_CACHE_MAX;
+
+/* dummy_len_plus values defined in ROM for SPI flash configuration */
+extern uint8_t g_rom_spiflash_dummy_len_plus[];
+static int extra_dummy = 0;
+typedef enum {
+    PSRAM_CMD_QPI,
+    PSRAM_CMD_SPI,
+} psram_cmd_mode_t;
+
+typedef struct {
+    uint16_t cmd;                /*!< Command value */
+    uint16_t cmdBitLen;          /*!< Command byte length*/
+    uint32_t *addr;              /*!< Point to address value*/
+    uint16_t addrBitLen;         /*!< Address byte length*/
+    uint32_t *txData;            /*!< Point to send data buffer*/
+    uint16_t txDataBitLen;       /*!< Send data byte length.*/
+    uint32_t *rxData;            /*!< Point to recevie data buffer*/
+    uint16_t rxDataBitLen;       /*!< Recevie Data byte length.*/
+    uint32_t dummyBitLen;
+} psram_cmd_t;
+
+static void IRAM_ATTR psram_cache_init(psram_cache_mode_t psram_cache_mode, psram_vaddr_mode_t vaddrmode);
+
+static void psram_clear_spi_fifo(psram_spi_num_t spi_num)
+{
+    int i;
+    for (i = 0; i < 16; i++) {
+        WRITE_PERI_REG(SPI_W0_REG(spi_num)+i*4, 0);
```

Comment:
Please add space between operand and operator.

Reply:
added. thanks